### PR TITLE
executor: refine the precision for avg

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
This commit ports the [old logic for handling decimal avg result](https://github.com/pingcap/tidb/blob/master/expression/aggregation/avg.go#L79-L91) .

### What is changed and how it works?
1. Set the `finalResult.RetTp.Decimal` to the result of type inferring when `appendFinalResult2Chunk` for `avg`.
2. Set the `finalResult.RetTp.Decimal` to 0 when the input argument is integer type for `sum`.

### Check List 

https://github.com/pingcap/tidb-test/pull/634

Code changes

 - Has exported function/method change

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: executor/aggfuncs/func_avg.go [hunk 1]
```diff
@@ -56,7 +58,19 @@ func (e *baseAvgDecimal) AppendFinalResult2Chunk(sctx sessionctx.Context, pr Par
 	finalResult := new(types.MyDecimal)
 	err := types.DecimalDiv(&p.sum, decimalCount, finalResult, types.DivFracIncr)
 	if err != nil {
-		return errors.Trace(err)
+		return err
+	}
+	// Make the decimal be the result of type inferring.
+	frac := e.args[0].GetType().Decimal
```
Comment (winoros) [inline executor/aggfuncs/func_avg.go]: Let it be a member of `baseAvgDecimal`?

File: executor/builder.go [hunk 2]
```diff
@@ -887,6 +887,27 @@ func (b *executorBuilder) wrapCastForAggArgs(funcs []*aggregation.AggFuncDesc) {
...
+			// that the `RetType` of `Column` refers to the `infoschema`, so we
+			// need to set a new variable for it to avoid modifying the
+			// definition in `infoschema`.
+			if col, ok := f.Args[i].(*expression.Column); ok {
+				col.RetType = types.NewFieldType(col.RetType.Tp)
+			}
+			// originTp is used when the the `Tp` of column is TypeFloat32 while
+			// the type of the aggregation function is TypeFloat64.
+			originTp := f.Args[i].GetType().Tp
+			*(f.Args[i].GetType()) = *(f.RetTp)
```
Comment (winoros) [inline executor/builder.go]: Should this thing be done during `typeinfer`

File: executor/aggfuncs/func_avg.go (hunk 1 above)
Comment (XuHuaiyu) [reply executor/aggfuncs/func_avg.go]: @winoros It seems we do not need to do this:
1. AppendFinalResultResult2Chunk will only be called **once** during the executing.
2. `avgOriginal4DistinctDecimal` does not hold `baseAvgDecimal` as a member variable, so we'll only use `frac` here even define it as a member of `baseAvgDecimal`.

File: executor/builder.go (hunk 2 above)
Comment (XuHuaiyu) [reply executor/builder.go]: @winoros Refer to https://github.com/pingcap/tidb/blob/master/executor/builder.go#L899-L903, we should do it here.

File: executor/aggfuncs/func_avg.go (hunk 1 above)
Comment (zz-jason) [inline executor/aggfuncs/func_avg.go]: Why not directly use the result type of the aggregate function, but use the result type of the argument instead?

Comment (XuHuaiyu) [reply executor/aggfuncs/func_avg.go]: @zz-jason 
We could not get the result type of function here except for adding an interface,
the argument type has been wrap with cast to make sure it's the same with the function.

**Comment by XuHuaiyu** [comment]
PTAL @winoros @zz-jason

**Comment by winoros** [comment]
BTW, do we need to change `SUM`'s?

**Comment by XuHuaiyu** [comment]
@winoros 
Maybe we should, I'll investigate it.
But it's better to push another to do this.

**Pull request merged**